File: pythoncodes/tsp.py
# Traveling salesman problem
# The first line indicates the number of cities. Each city is a point in the plane, 
# and each subsequent line indicates the x and y-coordinates of a single city.
# The distance between two cities is defined as the Euclidean distance, that is, two 
# cities at locations (x,y) and (z,w) have distance (xÅ|z)2+(yÅ|w)2 between them.
# In the box below, type in the minimum cost of a traveling salesman tour for this 
# instance, rounded down to the nearest integer. 

import math
import itertools

lines = [line.strip() for line in open('c:/home/algorithms2/tsp.txt')]
lines = [[float(num) for num in line.split(' ')] for line in lines]
numPoints = int(lines[0][0])
points = lines[1:]

def eucDist (x1, y1, x2, y2):
    "compute the euclidean distance between coordinates"
    return math.sqrt((x1 - x2)**2 + (y1 - y2)**2)

def tsp (coords):
    "Computes the shortest path between set of points, satisfying the condition that all the points must be visited once: the travelling salesman problem ( *O(n^2 * 2 ^n)* )"
    ## Initialize array with +inf except at S{1}
    A = 
    for m in range(2, numPoints):
        subsets = itertools.combinations(list(range(1,numPoints + 1)), m)
        subsWithOne = [sub for sub in subsets if any([p==1 for p in sub])]
        for sub in subsWithOne:
    pass

File: javacodes/TSP/src/tsp/TSP.java
package tsp;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;

public class TSP {
    static int numPoints;

    private static double eucDist(double x1, double x2, double y1, double y2) {
        return Math.sqrt(Math.pow((x1 - x2), 2) + Math.pow((y1 - y2), 2));
    }

    /**
     * Read in data: first line is number of vertices, Each subsequent line
     * describes a point (x y coordinates separated by a space)
     * 
     * @throws IOException
     */
    public static double[][] readIn(String filename) throws IOException {
        FileReader fileReader = new FileReader(filename);
        BufferedReader br = new BufferedReader(fileReader);
        String line = null;
        double[][] points = new double[1][1];
        int l = 0;
        int i = 0;
        while ((line = br.readLine()) != null) {
            String[] parts = line.split(" ");
            // Process the first line
            if (l == 0) {
                numPoints = Integer.parseInt(parts[0]);
                points = new double[numPoints][2];
                l++;
            } else {

                points[i][0] = Double.parseDouble(parts[0]);
                points[i][1] = Double.parseDouble(parts[1]);
                i++;
            }
        }
        br.close();
        return points;
    }

    /**
     * @param args
     * @throws IOException
     */
    public static void main(String[] args) throws IOException {
        double[][] points = readIn(args[0]);
        System.out.println("The number of points is " + numPoints);
        // for (int i = 0; i < numPoints; i++) {
        // System.out.println(points[i][0] + " " + points[i][1]);
        // }
        System.out.println(eucDist(points[0][0], points[0][1], points[1][0],
                points[1][1]));
    }
}
